akka-stream/src/main/scala/akka/stream/impl/io/IOSources.scala

```diff
@@ -4,46 +4,118 @@
-  extends SourceModule[ByteString, Future[IOResult]](shape) {
-  require(chunkSize > 0, "chunkSize must be greater than 0")
-  override def create(context: MaterializationContext) = {
-    // FIXME rewrite to be based on GraphStage rather than dangerous downcasts
-    val materializer = ActorMaterializerHelper.downcast(context.materializer)
-    val settings = materializer.effectiveSettings(context.effectiveAttributes)
+private[akka] object FileSource {
 
-    val ioResultPromise = Promise[IOResult]()
-    val props = FilePublisher.props(f, ioResultPromise, chunkSize, settings.initialInputBufferSize, settings.maxInputBufferSize)
-    val dispatcher = context.effectiveAttributes.get[Dispatcher](IODispatcher).dispatcher
+  val completionHandler = new CompletionHandler[Integer, Try[Int] ⇒ Unit] {
 
-    val ref = materializer.actorOf(context, props.withDispatcher(dispatcher))
+    override def completed(result: Integer, attachment: Try[Int] ⇒ Unit): Unit = {
+      attachment(Success(result))
+    }
 
-    (akka.stream.actor.ActorPublisher[ByteString](ref), ioResultPromise.future)
+    override def failed(ex: Throwable, attachment: Try[Int] ⇒ Unit): Unit = {
+      attachment(Failure(ex))
+    }
   }
+}
 
-  override protected def newInstance(shape: SourceShape[ByteString]): SourceModule[ByteString, Future[IOResult]] =
-    new FileSource(f, chunkSize, attributes, shape)
+/**
+ * INTERNAL API
+ * Creates simple asynchronous Source backed by the given file.
+ */
+private[akka] final class FileSource(path: Path, chunkSize: Int)
+  extends GraphStageWithMaterializedValue[SourceShape[ByteString], Future[IOResult]] {
+  require(chunkSize > 0, "chunkSize must be greater than 0")
 
-  override def withAttributes(attr: Attributes): Module =
-    new FileSource(f, chunkSize, attr, amendShape(attr))
+  val out = Outlet[ByteString]("FileSource.out")
+
+  override protected def initialAttributes: Attributes = DefaultAttributes.fileSource
+
+  override val shape = SourceShape(out)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[IOResult]) = {
+    val ioResultPromise = Promise[IOResult]()
+
+    val logic = new TimerGraphStageLogic(shape) {
+
+      val buffer = ByteBuffer.allocate(chunkSize)
+
+      var channel: AsynchronousFileChannel = _
+      var position = 0L
+      var chunkCallback: Try[Int] ⇒ Unit = _
+
+      override def preStart(): Unit = {
+        try {
+          // this is a bit weird but required to keep existing semantics
+          require(Files.exists(path), s"Path '$path' does not exist")
+          require(Files.isRegularFile(path), s"Path '$path' is not a regular file")
+          require(Files.isReadable(path), s"Missing read permission for '$path'")
+        } catch {
+          case ex: IllegalArgumentException ⇒
+            ioResultPromise.trySuccess(IOResult(position, Failure(ex)))
```

Comment:
Hmmm, I'm not sure that's all correct. 
If the file does not exist, should we not fail the entire future?